### ISSUE TITLE
[Salsa] Provide a JSDoc equivalent of `import { ..} from "mod"`

### ISSUE THREAD

**Issue by mhegazy** [Microsoft/TypeScript#14377]
Given an implementation of a class:

```ts
// file.js

export default class File {
   ...
}
```

How can you use the type without importing the module:

```ts
// another
export default Util {
   /** @param {File} file  */
   getFile(file) {
   }
}
```

One option is to use [`@module`](http://usejsdoc.org/tags-module.html)
e.g.:

```ts
   /** @param {module:./file:File} file  */
   getFile(file) {
   }
```

**Comment by bobzhang** [comment]
I hit the same issue today, the idea is to not relying on transpiling on server side, for example, I want to define my own object type like this, below does not work:

```ts
/**
 * @typedef {Object} Watcher
 * @property {string} dir
 * @property {fs.FSWatcher} watcher
 */
```

**Comment by mhegazy** [comment]
We will proceed with #14056  for now, try to special case classes.

**Comment by ste2425** [comment]
Such a workflow would also be brilliant for Angular1 apps where you may have a main file that injects all your dependencies and a separate file that specifies your controller. The controller file exports a function/class. It receives its dependencies but does not inject them directly, being able to tag them using JSdoc as a type that was defined in a different file in your workspace would be invaluable there.

**Comment by drKnoxy** [comment]
@mhegazy did you ever get anything working along these lines? Trying to reference a js class from a typedef, sounds like the same problem.

**Comment by goodmind** [comment]
Any progress here?

**Comment by LestaD** [comment]
Maybe?

```js
/**
 * @import {File} './file'
 */

/**
 * @param {File} file
 */
getFile(file) {
}
```

**Comment by Kuzirashi** [comment]
Helo, was there any update on this? It would be great to have this.

**Comment by realappie** [comment]
This would be a great addition to vscode!!

**Comment by sofib** [comment]
Having a straightforward way to do this is a must in any decent code design where one needs to inject a dependency. Any updates, please?

**Comment by hitendramalviya** [comment]
Any update on this?

**Comment by clshortfuse** [comment]
I suggest we use the same syntax as Closure does:

```
/** @param {./foo.Foo} Foo */
function(Foo) {}
```

https://github.com/google/closure-compiler/wiki/JS-Modules#type-references

You can also write the syntax as `/** @type {./path/to/module.propName} */`

**Comment by crabicode** [comment]
Sadly, the conversation has not gone far to implement this feature:( Very much looking forward to having this added.

**Comment by joma74** [comment]
I suggest the use of the same syntax and ts mechanism as es6 does:
/// import Vue from "vue";

**Comment by DanielRosenwasser** [comment]
We're going to hold off on any new JSDoc syntax, and wait for more feedback here. As per #22445, our current recommendation is to wait on #14844 and then use

```js
/**
 * @typedef {import("express")} express
 */
```

**Comment by mhegazy** [comment]
closing in favor of #14844

**Comment by clshortfuse** [comment]
@DanielRosenwasser 

Just remember that Salsa also works with ES2016/ES6 class imports, not just TypeScript. That syntax looks rather foreign to me since I don't use TS.

**Comment by mhegazy** [comment]
The syntax in the proposal is not a TS syntax either. we have surveyed github code bases and did not find a common syntax to use instead. use of JSDoc `/* @module */` or `/* @exports */` was very sparse, if non-existent. We picked `import` since it looks like ES6 module imports.

**Comment by joma74** [comment]
Is this covered somewhere after #14844 has been closed? What has been picked? Please plan forward to  document it at [https://github.com/Microsoft/TypeScript/wiki/JsDoc-support-in-JavaScript] or [https://www.typescriptlang.org/docs/handbook/type-checking-javascript-files.html]

**Comment by mhegazy** [comment]
The issue has been closed in favor of https://github.com/Microsoft/TypeScript/issues/14844, which is fixed by https://github.com/Microsoft/TypeScript/pull/22592. The change should be included in the next release TypeScript 2.9. 
with that you should be able to write:

```ts
/** @type {import("./externs").Bar} */
var a;
```
or 
```ts
/** @type {typeof import("./externs").Bar}
var b;
```
More docs should be available as we approach the release date for TS 2.9

**Comment by ipengyo** [comment]
no work

![image](https://user-images.githubusercontent.com/9323722/39088624-0bbf5914-45e8-11e8-8793-18f31011dad9.png)


version  1.22.2 (1.22.2)

**Comment by mhegazy** [comment]
Should be working correctly in latest drop of TypeSript (@next). VSCode ships with an older version of TypeScript. Please see [Using Newer TypeScript Versions](https://code.visualstudio.com/docs/languages/typescript#_using-newer-typescript-versions) documentation for more details on updating your VSCode to use a different version of TypeScript.

**Comment by mhegazy** [comment]
use `typeof` instead:

```ts
/** @typedef {typeof import('http')} http */
```

**Comment by thw0rted** [comment]
Since #14844 doesn't mention JSDoc at all, I'm asking here -- how does this new functionality impact eslint?  If I get this working in `tsc` and/or VS Code, will I still be able to run "vanilla" eslint and have it understand the updated syntax?  If not, is "this breaks eslint" a good enough reason to reopen this ticket or file a new one?

**Comment by mhegazy** [comment]
looks like a question/suggestion for the eslint issue tracker.

**Comment by thw0rted** [comment]
I mentioned it [on an eslint issue](https://github.com/eslint/eslint/issues/9050) and it sounds like it's fine -- I was just worried that maybe TS was making a vendor-specific JSDoc extension / syntax that wouldn't be compatible with other tools.

**Comment by yulianovdey** [comment]
I was following this issue assuming that the change would allow IntelliSense suggestions/parameter info to work inside a `.js` file without doing an explicit `require` for the module, even for files that don't have typings. I've installed `typescript@next` and pointed to it using `typescript.tsdk`, but I can't get suggestions to work doing something like this:

```
/** @typedef {import('./myService')} myService */

/**
 * @param {myService} myService
 */
function myFunction(myService) {
  myService. // Doesn't show the same suggestions that require('./myService') shows
}
```

Is there a way to get this to work, or is it separate from this issue?

**Comment by mhegazy** [comment]
what does `myService.js` exports look like?

**Comment by yulianovdey** [comment]
Thanks for the quick reply! Just some functions with doc blocks:

```
/**
 * @param {string} param1
 * @param {string} param2
 */
function moduleFunc(param1, param2) {

}
module.exports = {
  moduleFunc,
};
```

**Comment by mhegazy** [comment]
@yulianovdey, filed https://github.com/Microsoft/TypeScript/issues/24064 to track this issue.

**Comment by yulianovdey** [comment]
@mhegazy Thanks! I appreciate it.

**Comment by joma74** [comment]
Is/Can the following case be covered/coverable?

type of `webpack.Configuration` defines `devServer` as `any`.  So would like to apply type of `webpack-dev-server.Configuration` to the `devServer` property.

In the following code devServer is still identified as `(property) webpack.Configuration.devServer?: any`. 

```
/**
 * @type {import ("webpack").Configuration}
 */
var webpackConfig =
  {
    /**
     * @type {import("webpack-dev-server").Configuration}
     */
    devServer: {
      port: 8080,
      watchContentBase: false,
      hot: true,
      stats: "errors-only",
      host: "0.0.0.0",
    },
```

**Comment by thw0rted** [comment]
You could put in a PR against DefinitelyTyped to fix `@types/webpack`, then you wouldn't have to worry about it.  In fact, the typings file I'm looking at even has "// TODO: Type this" on `devServer` :D

**Comment by joma74** [comment]
Thanks for the hint;
Still I think that my more nearer/specific typing may/should be applied. On the general case that artifacts are meant to be not coupled.

**Comment by vbogach** [comment]
Unfortunately, this code breaks my jsdoc generation:
```javascript
/**
 * @typedef {typeof import("./foo").Bar} Bar
 */
```
Results into:

> ERROR: Unable to parse a tag's type expression for source file in line 1 with tag title "typedef" and text "{typeof import("./foo").Bar} Bar": Invalid type expression "typeof import("./foo").Bar": Expected "|" but "i" found.

when I execute jsdoc. Could you give any hints on how to keep my jsdoc working?

**Comment by mhegazy** [comment]
looks like a request for your jsdoc generation tool.

**Comment by Thaina** [comment]
@mhegazy Does not sem like that work

![image](https://user-images.githubusercontent.com/1042507/41705262-b0dc5128-7562-11e8-9306-3e9f71b358f4.png)

Currently now only this below work

![image](https://user-images.githubusercontent.com/1042507/41705344-e768d888-7562-11e8-8f73-dadfc45e8467.png)

**Comment by mhegazy** [comment]
you have declared a type alias called `firebase` whose type is `typeof import('firebase')`, but that has nothing to do with the variable `firebase` that is not defined. 

consider adding a .d.ts file in your project with a declaration for `firebase`:

```ts
// globals.d.ts

declare var firebase: typeof import("firebase");
```

also `/* global ... */` is not supported at the moment, https://github.com/Microsoft/TypeScript/issues/15626 tracks adding that.

**Comment by Thaina** [comment]
@mhegazy It is still the same that it need to have d.ts separate from js while I would like to declare in js purely (or don't even need to declare anything if possible)

ps. The `/* global */` there is not related. I just put it to satisfied eslint

**Comment by Thaina** [comment]
@mhegazy Also if I put the `declare var` in some d.ts into the project. I need to actively open it in the tab. Else it was not counted in the engine

It work fine here

![image](https://user-images.githubusercontent.com/1042507/41832439-3da06f34-7876-11e8-9bfd-7d87588ec738.png)

Until I close the file

![image](https://user-images.githubusercontent.com/1042507/41832429-35750586-7876-11e8-89b0-003828dcb3bb.png)

**Comment by mhegazy** [comment]
> Also if I put the declare var in some d.ts into the project. I need to actively open it in the tab. Else it was not counted in the engine

Make sure you add a `jsconfig.json` at the root of your project.

**Comment by niedzielski** [comment]
> Unfortunately, this code breaks my jsdoc generation:
> 
> ```js
> /**
>  * @typedef {import("./foo").Bar} Bar
>  */
> ```
This issue is encompassed in https://github.com/jsdoc3/jsdoc/issues/1537.

[JSDoc is not a very active project](https://github.com/jsdoc3/jsdoc/graphs/commit-activity) so I think it's unlikely to add support for this functionality. Unfortunately, this makes it very difficult to migrate vanilla JavaScript to TypeScript type checking while still appeasing JSDoc stakeholders wary of change since the new syntax breaks JSDoc execution. I've experimented with `/// <reference path="foo.js" />` syntax but haven't been able to find a substitute that will allow TypeScript to pull in the JSDoc typing.

If it were supported, the `/// import {Foo} from './bar'` syntax proposed by @joma74 or even `/** @import {Foo} from './bar' */` by @sergeysova would work for most folks since [the default JSDoc config](http://usejsdoc.org/about-configuring-jsdoc.html#default-configuration-options) allows unknown tags such as `@import`.

**Comment by SystemDisc** [comment]
@joma74 You could actually do the following:
```typescript
/**
 * @type {import ("webpack").Configuration & { devServer: import("webpack-dev-server").Configuration }}
 */
var webpackConfig = {
  devServer: {
    port: 8080,
    watchContentBase: false,
    hot: true,
    stats: "errors-only",
    host: "0.0.0.0",
  },
  // ...
};
```

**Comment by Thaina** [comment]
@mhegazy 

> declare var firebase: typeof import("firebase");

I made this file in project and it was add to `.d.ts` and normal intellisense work correctly

![image](https://user-images.githubusercontent.com/1042507/42752632-b96bfe96-8918-11e8-832d-a5ea36cbee5c.png)

However the JSDoc does not register the class in namespace loaded this way

![image](https://user-images.githubusercontent.com/1042507/42752735-0ffa80fc-8919-11e8-812d-40a618e9b6f1.png)

**Comment by mhegazy** [comment]
Can you share a project i can look at?

**Comment by Thaina** [comment]
@mhegazy Thank you

I was make it compact so might need to run `npm install` first

[TTTT.zip](https://github.com/Microsoft/TypeScript/files/2200461/TTTT.zip)

**Comment by nemoDreamer** [comment]
I'm surprised that this was closed as "fixed", when the intent was to "provide a JSDoc equivalent of ...".

`/** @typedef {import('foo').Bar} Bar */` is **not** a "JSDoc equivalent", since it's invalid JSDoc...!

As already pointed out by https://github.com/Microsoft/TypeScript/issues/14377#issuecomment-400464783, it would be great to have a **real** equivalent that could would _simultaneously_ let Typescript-in-Javascript know about the `Bar` type _and_ not have JSDoc complain about not knowing `Bar` in regular `@type {Bar}`/`@param {Bar} bar` usage throughout the file.

❤️

**Comment by vschoener** [comment]
Sadly, we still have nothing good to use? :(

**Comment by SystemDisc** [comment]
It would be great if this issue could be reopened to continue discussion and not let it get lost.

**Comment by saschanaz** [comment]
Do even JSDoc have an official syntax for type import?

**Comment by maxjf1** [comment]
Is there an way to reference an class as an argument without importing it?

**Comment by ricardohbin** [comment]
I did a little patch in JSDoc to does not breaks and accepts `{import('./file').Foo}` -> https://github.com/ricardohbin/jsdoc/pull/1
